src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1724,104 +1726,294 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
+        /// Note that the variables may either be plain or nested variables.
+        /// Returns false if there was an error.
+        /// </summary>
+        private bool DeconstructIntoSteps(
+                        BoundDeconstructValuePlaceholder targetPlaceholder,
+                        AssignmentExpressionSyntax syntax,
+                        DiagnosticBag diagnostics,
+                        ImmutableArray<BoundExpression> variables,
+                        ArrayBuilder<DeconstructStep> deconstructionSteps,
+                        ArrayBuilder<AssignmentStep> assignmentSteps)
+        {
+            Debug.Assert(targetPlaceholder.Type != null);
 
-            // tuple literal such as `(1, 2)`, `(null, null)`, `(x.P, y.M())`
-            if (boundRHS.Kind == BoundKind.TupleLiteral || ((object)boundRHS.Type != null && boundRHS.Type.IsTupleType))
+            DeconstructStep step;
+
+            if (targetPlaceholder.Type.IsTupleType)
+            {
+                // tuple literal such as `(1, 2)`, `(null, null)`, `(x.P, y.M())`
+                step = MakeTupleDeconstructStep(targetPlaceholder, syntax, diagnostics, variables, deconstructionSteps, assignmentSteps);
+            }
+            else
             {
-                return BindDeconstructWithTuple(node, diagnostics, boundRHS, checkedVariables);
+                step = MakeNonTupleDeconstructStep(targetPlaceholder, syntax, diagnostics, variables, deconstructionSteps, assignmentSteps);
             }
 
-            // expression without type such as `null`
-            if (boundRHS.Type == null)
+            if (step == null)
             {
-                Error(diagnostics, ErrorCode.ERR_DeconstructRequiresExpression, node);
-                return BadExpression(node, checkedVariables.Concat(boundRHS).ToArray());
+                return false;
             }
 
-            return BindDeconstructWithDeconstruct(node, diagnostics, boundRHS, checkedVariables);
+            deconstructionSteps.Add(step);
+
+            // outputs will either need a conversion step and assignment step, or if they are nested variables, they will need further deconstruction
+            return DeconstructOrAssignOutputs(step, variables, syntax, diagnostics, deconstructionSteps, assignmentSteps);
         }
 
-        private BoundExpression BindDeconstructWithTuple(AssignmentExpressionSyntax node, DiagnosticBag diagnostics, BoundExpression boundRHS, ImmutableArray<BoundExpression> checkedVariables)
+        /// <summary>
+        /// This will generate and stack appropriate deconstruction and assignment steps for a tuple type.
+        /// The produced deconstruction step has no Deconstruct method since the tuple already has distinct elements.
+        /// </summary>
+        static private DeconstructStep MakeTupleDeconstructStep(
+                        BoundDeconstructValuePlaceholder targetPlaceholder,
+                        AssignmentExpressionSyntax syntax,
```

Comment:
The syntax could easily be tracked through syntax that BoundDeconstructionVariables nodes are associated with.

---
In reply to: [65782761](https://github.com/dotnet/roslyn/pull/11715#discussion_r65782761) [](ancestors = 65782761)